feat(backend): add window size api

benjajaja opened the pull request:
Depends on https://github.com/crossterm-rs/crossterm/pull/790, and should update crossterm in an independent PR. Related to #215

For image (sixel, iTerm2, Kitty...) support that handles graphics in
terms of `Rect` so that the image area can be included in layouts.

For example: an image is loaded with a known pixel-size, and drawn, but
the image protocol has no mechanism of knowing the actual cell/character
area that been drawn on. It is then impossible to skip overdrawing the
area.

Returning the window size in pixel-width / pixel-height, together with
colums / rows, it can be possible to account the pixel size of each cell
/ character, and then known the `Rect` of a given image, and also resize
the image so that it fits exactly in a `Rect` and avoid artifacts.

![image](https://github.com/tui-rs-revival/ratatui/assets/310215/6271814b-3d4f-4cd0-ad9e-f810e54b9f59)

joshka commented on Cargo.toml:
```diff
@@ -37,7 +37,8 @@ cargo-args = ["-Zunstable-options", "-Zrustdoc-scrape-examples"]
 bitflags = "2.3"
 cassowary = "0.3"
-crossterm = { version = "0.26", optional = true }
+# crossterm = { version = "0.26", optional = true }
+crossterm = { git = "https://github.com/benjajaja/crossterm", rev = "ad84a3829d55eac27cfdb7791b2e34c1732040ad", optional = true }
```
I think I'd prefer to wait until this is merged to merge this into ratatui. This is fine as a draft for now however.

You can also set this in a patch section and leave the main dependency in place. See https://doc.rust-lang.org/cargo/reference/overriding-dependencies.html#the-patch-section

```toml
[patch.crates-io]
crossterm = { git = "https://github.com/benjajaja/crossterm", rev = "ad84a3829d55eac27cfdb7791b2e34c1732040ad", optional = true }
```

benjajaja replied:
Oooh TIL, that's so cool, thank you!

joshka commented on examples/window_size.rs:
```diff
@@ -0,0 +1,80 @@
+use std::{error::Error, io};
```
It could be worthwhile adding some doc comments to help a reader understand what this example shows.

benjajaja replied:
I removed this in the end, as it's not a valuable demo/example.

joshka commented on src/backend/crossterm.rs:
```diff
@@ -86,7 +86,7 @@ where
             // Move the cursor if the previous location was not (x - 1, y)
             if !matches!(last_pos, Some(p) if x == p.0 + 1 && y == p.1) {
-                map_error(queue!(self.buffer, MoveTo(x, y)))?;
+                queue!(self.buffer, MoveTo(x, y))?;
```
I see this is removed a bunch of times. Can you please find out or explain why this was necessary in the first place?

benjajaja replied:
It seems like the `queue` macro, or `QueueableCommand::queue`, now return `io::Result` making the `map_error` unnecessary.

joshka commented on src/backend/mod.rs:
```diff
@@ -112,6 +112,10 @@ pub trait Backend {
     /// Get the size of the terminal screen as a [`Rect`].
     fn size(&self) -> Result<Rect, io::Error>;
 
+    /// Get the size of the terminal screen as well as windows sixe in pixels as a [`Rect`] and a
+    /// `(u16, u16)` tuple.
+    fn window_size(&mut self) -> Result<(Rect, (u16, u16)), io::Error>;
```
Do you think it makes sense to split this to two methods with more explicit names instead of a single returning a tuple? Is the Rect ever not positioned at x=0, y=0?

benjajaja replied:
I say no, because 2 out of 3 implementations return both things at once in one call, due to the underlying API.

benjajaja replied:
Oh and the `Rect` is just like `size`, always x=0,y=0.

joshka commented on src/backend/termwiz.rs:
```diff
@@ -221,3 +220,18 @@ impl From<Color> for ColorAttribute {
+    }
+}
+
+fn u16_max(i: usize) -> u16 {
+    let max = u16::max_value();
+    if i > usize::from(max) {
+        max
+    } else {
+        i as u16
+    }
+}
```
Replace with `u16::try_from(i).unwrap_or(u16::MAX);` (probably obvious enough to just inline into the methods that use it)

benjajaja replied:
Much better, thank you!

joshka commented on src/backend/test.rs:
```diff
@@ -177,6 +177,10 @@ impl Backend for TestBackend {
     }
 
+    fn window_size(&mut self) -> Result<(Rect, (u16, u16)), io::Error> {
+        Ok((Rect::new(0, 0, self.width, self.height), (7, 12)))
```
What's the significance of 7 and 12 here? Can you document it?

benjajaja replied:
It was from when it was `font_size`, not `window_size` - my bad.
Updated to a better value and added a comment.

joshka reviewed (changes requested):
Rewrite the commit message title in the convention commit style using imperative tone (e.g. `feat(backend): add window_size_in_pixels() method`).

This is pretty close to being mergable (or will be once the crossterm PR is merged), so can you please rebase and squash your commits into a single commit and setup commit signing? See https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits for info.

joshka commented:
Backend is woefully under-tested. I wonder if there's some way to fix that.

benjajaja commented:
Regarding tests, I don't think it's really feasible to cover this. `window_size()` really depends on being run as a graphical application, i.e. xorg or wayland.

joshka reviewed (changes requested):
I think there was a mention of this not yet working with windows on crossterm if I read the linked issue right. Can you make sure to document this in the crossterm docs for the window_size method?

joshka commented on src/backend/mod.rs:
```diff
@@ -112,6 +112,9 @@ pub trait Backend {
     fn size(&self) -> Result<Rect, io::Error>;
 
+    /// Get the font size of the terminal screen as [`(width, height)`].
+    fn window_size(&mut self) -> Result<(Rect, (u16, u16)), io::Error>;
```
I've been thinking about this a little - I think there's a few places within various widgets where we care more about the particular offset that a size. A good example of that `Buffer::empty()` which takes a Rect, but really only needs that to be a Size. We use this a lot for tests. So we might treat a Rect as a Point(x,y) and a Size(width, height).

It's probably out of scope for this PR to do all that however. PixelSize is still a bit ambiguous (I could see users getting confused over whether it is the size of a cell in pixels or the size of a window?)

What about:
```rust
struct WindowSize {
  size_in_rows_and_cols: Size,
  size_in_pixels: Size,
}
```

> Changing the return type of size() would also mean that this is a breaking change. Perhaps it would be better to have window_size() just return (u16, u16) as you mentioned? What if that was named `size_in_pixels()`?

joshka commented:
BTW, another PR upgraded crossterm 0.27 and brought in the changes to the error bits. Can you rebase this on main please?

benjajaja commented on src/terminal.rs:
```diff
@@ -420,7 +420,7 @@ where
         self.backend.append_lines(height)?;
         let missing_lines =
-            height.saturating_sub(self.last_known_size.bottom() - self.viewport_area.top());
+            height.saturating_sub(self.last_known_size.height - self.viewport_area.top());
```
`last_known_size` was a `Rect` that was only set from `with_options()` and `resize()`. Both did only have non-zero x,y for `Viewport::Fixed`. This method only runs for `Viewport::Inline(_)` and it is the only method that uses the field.

benjajaja commented:
I tried adding some test coverage for `insert_before`, but that did require to implement `clear_region()` and `append_line()` for `TestBackend` for which I don't know exactly how to implement the different "clear types", e.g. `AfterCursor` etc. In general, it seemed a bit pointless to test `insert_before` against `TestBackend`.

joshka commented on src/backend/mod.rs:
```diff
@@ -109,8 +114,16 @@ pub trait Backend {
         }
     }
 
-    /// Get the size of the terminal screen as a [`Rect`].
-    fn size(&self) -> Result<Rect, io::Error>;
+    /// Get the size of the terminal screen in columns/rows as a [`Rect`].
+    fn size(&self) -> Result<Size, io::Error>;
```
How about leaving this as Rect and just updating the window_size method?

otherwise nit: doc comment => Size instead of Rect

joshka commented on src/layout.rs:
```diff
@@ -587,6 +587,41 @@ fn try_split(area: Rect, layout: &Layout) -> Result<Rc<[Rect]>, AddConstraintErr
+}
+
+impl Size {
+    /// Creates a new size, with width and height limited to keep the area under max u16.
+    /// If clipped, aspect ratio will be preserved.
```
Do we actually need this truncation for anything?

joshka reviewed (changes requested):
I think perhaps make this non-breaking with the existing size() and resize() continuing to take Rects? We can come back to that decision at a later date however.

Aiming for a 0.23.0 release after the weekend, so this should be in that if you can make the changes in time.

benjajaja commented:
@joshka I toned this down to only adding the new `window_resize()` method, still introducing the new `Size` struct but not using it in src/terminal.rs so no breaking changes. Also dropped the truncation in `Size` copied from `Rect`, as its main purpose is to keep `area` (and maybe indirectly `Buffer::content`) in bounds, and there are no more conversions between `Size` and `Rect` now.

It's cool if we can get this into the release, but it's not critical like the cell skipping, as the window pixel size can be queried outside of `ratatui`.

--------------
Some thoughts on `Size` vs `Rect` in `Terminal`:

In `Terminal`, there is a `last_known_size` that may have non-zero x,y if the `viewport` is `Viewport::Fixed(Rect)`. However it seems like the field is used exclusively by `Viewport::Inline(u16)`.

In general, I think it could be healthy to cut down a few public methods in `Terminal`: `autoresize`, `resize`, `size` (is just a wrapper to `backend.size()`), and `area` of `CompletedFrame` returned by `draw`.

joshka commented on src/backend/mod.rs:
```diff
@@ -60,6 +60,11 @@ pub enum ClearType {
 
+pub struct WindowSize {
+    pub columns_rows: Size,
+    pub window_pixels: Size,
```
nit: what do you think about naming this just `pixels` given that the struct is named **`Window`**`Size`?

kdheepak replied:
I think there should be a doc string here too, right? Maybe describing when someone might want to use this.

joshka reviewed (approved):
This looks good now. Thanks again. This seems like something we could merge before release. I'll flip it to the other maintainers to take a look.

Comments are non-blocking

mindoodoo reviewed (approved):
Looks good ! Thanks for this !

I'll have to take a look at the image crate soon ! Exciting stuff.

@joshka what do you think about the scope of the conventional commit PR title ? let me know if anything else would be more suited. Thanks for all the review on this !

joshka commented:
> @joshka what do you think about the scope of the conventional commit PR title ? let me know if anything else would be more suited. Thanks for all the review on this !

I think it's probably ok as is. Merging this now. Thanks again @benjajaja. This is going to lead to some fun apps.
